### PR TITLE
The task of creating a universal storage class has been solved.

### PR DESCRIPTION
**Pull request by 4Vitalii5**
method addKeyPair, addValueByKey has been created. The logic of the put, get, size methods has been implemented.

### CODE REVIEW

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 1]
```diff
@@ -3,17 +3,64 @@
 import core.basesyntax.Storage;
 
 public class StorageImpl<K, V> implements Storage<K, V> {
+    private static final int MAX_ARRAY_LENGTH = 10;
+    private final Object[] keyArray;
+    private final Object[] valueArray;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:8]: use K,V

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 2]
```diff
@@ -3,17 +3,64 @@
...
+    private int arraySize = 0;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:9]: 0 is default for int

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 3]
```diff
@@ -3,17 +3,64 @@
...
+        keyArray = new Object[MAX_ARRAY_LENGTH];
+        valueArray = new Object[MAX_ARRAY_LENGTH];
+    }
+
     @Override
     public void put(K key, V value) {
+        if (arraySize >= MAX_ARRAY_LENGTH) {
+            throw new IllegalStateException("Storage is full. Maximum size is "
+                    + MAX_ARRAY_LENGTH + " .");
+        } else if (!updateValueByKey(key, value)) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:21]:
```suggestion
        } 
        if (!updateValueByKey(key, value)) {
```
we don't need else after returning/throwing

**Review by okuzan — CHANGES_REQUESTED** [review]
please use throughout your code correct comparison condition - a == b or a != null  and a.equals(b)

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 4]
```diff
@@ -3,17 +3,65 @@
...
+
     @Override
     public void put(K key, V value) {
+        if (arraySize >= MAX_ARRAY_LENGTH) {
+            throw new IllegalStateException("Storage is full. Maximum size is "
+                    + MAX_ARRAY_LENGTH + " .");
+        }
+        if (!updateValueByKey(key, value)) {
+            addKeyPair(key, value);
+        }
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:24]: you can combine in one method and make it more simple solution

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 5]
```diff
@@ -3,17 +3,65 @@
...
     }
 
     @Override
     public V get(K key) {
+        for (int i = 0; i < keyArray.length; i++) {
+            if (keyArray[i] != null && keyArray[i].equals(key)) {
+                return valueArray[i];
+            } else if (keyArray[i] == key && valueArray[i] != null) {
+                return valueArray[i];
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:33]: after you can reuse the solution of method which you will use for method put

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 6]
```diff
@@ -3,17 +3,54 @@
 import core.basesyntax.Storage;
 
 public class StorageImpl<K, V> implements Storage<K, V> {
+    private static final int MAX_ARRAY_LENGTH = 10;
+    private final K[] keyArray;
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:7]:
```suggestion
    private final K[] keys;
```

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 7]
```diff
@@ -3,17 +3,54 @@
...
+    private final V[] valueArray;
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:8]:
```suggestion
    private final V[] values;
```

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 8]
```diff
@@ -3,17 +3,54 @@
...
+        valueArray = (V[]) new Object[MAX_ARRAY_LENGTH];
+    }
+
     @Override
     public void put(K key, V value) {
+        if (arraySize >= MAX_ARRAY_LENGTH) {
+            throw new IllegalStateException("Storage is full. Maximum size is "
+                    + MAX_ARRAY_LENGTH + " .");
+        }
+        saveOrUpdate(key, value);
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:22]:
```suggestion
        int index = indexOf(key);
        if (index != -1) {
            values[index] = value;
        } else if (index == -1) {
            keys[size] = key;
            values[size] = value;
            size++;
        }
```

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 9]
```diff
@@ -3,17 +3,54 @@
...
     }
 
     @Override
     public V get(K key) {
+        for (int i = 0; i < keyArray.length; i++) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:27]:
```suggestion
        int index = indexOf(key);
        return index == -1 ? null : values[index];
```

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 10]
```diff
@@ -3,17 +3,54 @@
 import core.basesyntax.Storage;
 
 public class StorageImpl<K, V> implements Storage<K, V> {
+    private static final int MAX_ARRAY_LENGTH = 10;
+    private final K[] keyArray;
+    private final V[] valueArray;
+    private int arraySize;
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:9]:
```suggestion
    private int size;
```

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 11]
```diff
@@ -3,17 +3,54 @@
...
         return null;
     }
 
     @Override
     public int size() {
-        return -1;
+        return arraySize;
+    }
+
+    private void saveOrUpdate(K key, V value) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:42]:
```suggestion
    private int indexOf(K key) {
        for (int i = 0; i < size; i++) {
            if (key == keys[i] || key != null && key.equals(keys[i])) {
                return i;
            }
        }
        return -1;
    }
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request closed without merging**